opacus/grad_sample/grad_sample_module.py

```diff
@@ -318,7 +322,6 @@ def rearrange_grad_samples(
                 " run forward after add_hooks(model)"
             )
 
-        # TODO: that's not right; DPLSTM isn't always batch_first=False
         batch_dim = 0 if batch_first or type(module) is RNNLinear else 1
```

Comment:
I just found that this line does something hacky. A don't understand why we propagate batch_first to all submodules in `add_hooks()`, is this what we want to do?

Why `RNNLinear` is the only exception here? Does this work correctly with `nn.Linear`? I have a feeling that this is a bug not covered by tests.

The method `add_hooks()` goes recursively through all submodules and set the global `batch_first`:
```python
        for module in trainable_modules(self._module):
            if type(module) in self.GRAD_SAMPLERS:
                ...
                self.autograd_grad_sample_hooks.append(
                    module.register_backward_hook(
                        partial(
                            self.capture_backprops_hook,
                            ...
                            batch_first=batch_first,
                        )
                    )
                )
```

I found this behavious counter-intuitive because the module can easily have submodules with different batch_dim and dim rearranges in between. For example, in `DPRNN`:
```
model: DPRNN(batch_dim=1)
  - cell: nn.Linear(batch_dim=0)
```
In case of DPRNN (as well as the original nn.RNN) batch rearranging happens inside the implementation. 

If not that `or type(module) is RNNLinear` we would try to treat nn.Linear as it uses `batch_dim=1` which is not the case, it alwayse uses batch_dim=0.

Reply:
Based on what you described, I don't think we have a clear bug. 
But we certainly have an implicit assumption about the model, which we should communicate better. (If you have any ideas on how to verify / enforce it, they'll be very welcome)

In addition to everything we check in `ModelValidator`, we also assume the model doesn't do any weird reshapes. On all intermediate stages we expect the model to maintain batch_dimension and either first or second dimension. (and yes, we don't verify that). 
On a high level, it's the same as us prohibiting `BatchNorm` - samples must be processed independently of each other, and the output of any batch must be equivalent to micro-brach processing (process each sample individually with batch_size=1 and average the results). It's not the case for `BatchNorm`, and it's also not the case if you do custom tensor operations and don't preserve batch dimension.

> A don't understand why we propagate batch_first to all submodules in add_hooks(), is this what we want to do?

In the current implementation - yes, because of the implicit assumption above.

> Why RNNLinear is the only exception here? Does this work correctly with nn.Linear?

Because we assume `RNNLinear` to be only used in `DPRNN`, and we know `DPRNN` does this with batch dimension.
It's also the reason we can't get rid if `RNNLinear` in favour of `nn.Linear` just yet

Reply:
Thanks for explaining me. Wow, this is a strong assumption, it's definitely worth to make it clear.

Although, I'm still trying to understand how it can work.

> On all intermediate stages we expect the model to maintain batch_dimension and either first or second dimension. (and yes, we don't verify that).

What happens to the inner nn.Linear when we use batch_dim=1? 
This hack with RNNLinear looks to me as a solution of this problem, but how other modules work?
For example, in `DPMultiheadAttention` is always batch_dim=1, but there are nn.Linear layers which always batch_dim=0.